GVFS/GVFS.Common/Cleanup/GitCleanupStep.cs

```diff
@@ -0,0 +1,198 @@
+                    }
+                }
+                else
+                {
+                    this.CreateProcessAndRun();
+                }
+            }
+            catch (IOException e)
+            {
+                this.Context.Tracer.RelatedWarning(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.TelemetryKey + ": IOException while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+            }
+            catch (Exception e)
+            {
+                this.Context.Tracer.RelatedError(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.TelemetryKey + ": Exception while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+                Environment.Exit((int)ReturnCode.GenericError);
+            }
+        }
+
+        public void Stop()
+        {
+            lock (this.gitProcessLock)
+            {
+                this.Stopping = true;
+
+                GitProcess process = this.GitProcess;
+
+                if (process != null)
+                {
+                    if (process.TryKillRunningProcess())
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.TelemetryKey + ": killed background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                    else
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.TelemetryKey + ": failed to kill background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Implement this method to perform a cleanup step. If the object-cache lock is required
+        /// (as specified by <see cref="RequireCacheLock"/>), then this step is not run unless we
+        /// hold the lock.
+        /// </summary>
+        protected abstract void RunGitAction();
+        
+        protected GitProcess.Result RunGitCommand(Func<GitProcess, GitProcess.Result> work)
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity("RunGitCommand", EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                if (this.Stopping)
+                {
+                    this.Context.Tracer.RelatedWarning(
+                        metadata: null,
+                        message: this.TelemetryKey + ": Not launching Git process because the mount is stopping",
+                        keywords: Keywords.Telemetry);
+                    return null;
+                }
+
+                GitProcess.Result result = work.Invoke(this.GitProcess);

```

Comment:
Does this need the `gitProcessLock`?  If not I'm a little confused why we need the lock at all.

Reply:
The process lock is for changing the state of the process (initializing it, calling `TryKillRunningProcess()` during stop). It prevents us from creating a new process in `CreateProcessAndRun()` if we are stopping.

Locking the process during the execution of a git command would stop us from being able to halt the process.